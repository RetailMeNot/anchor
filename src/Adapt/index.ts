export { Adapt } from './Adapt.component';
