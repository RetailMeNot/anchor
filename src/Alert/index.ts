export { Alert } from './Alert.component';
