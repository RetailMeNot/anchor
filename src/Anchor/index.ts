export { default as Anchor } from './Anchor.component';
