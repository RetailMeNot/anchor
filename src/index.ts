export { Avatar } from './Avatar';
export { AutoComplete } from './AutoComplete';
export { Button } from './Button/Button.component';
export { Badge } from './Badge';
export { DropDown } from './DropDown';
export { Container, Flex, Box } from './Grid';
export { Icon } from './Icon';
export { Page, Footer, Header } from './Layout';
export { List, Item as ListItem } from './List';
export { Menu, Item as MenuItem } from './Menu';
export { colors, fonts, sizes, variables, GlobalCSS, NormalizeCSS } from './theme';
