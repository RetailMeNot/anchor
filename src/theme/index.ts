export { default as colors } from './colors.theme';
export { default as fonts } from './fonts.theme';
export { default as sizes } from './sizes.theme';
export { default as variables } from './variables.theme';

export { NormalizeCSS, GlobalCSS } from './GlobalStyles';
