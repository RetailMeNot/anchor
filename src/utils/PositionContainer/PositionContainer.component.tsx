// VENDOR
import * as React from 'react';
import { forwardRef } from 'react';
import styled, { css } from '@xstyled/styled-components';
import { th } from '@xstyled/system';
import classNames from 'classnames';
// COMPONENTS
import { positionVariants, Position } from '../position/position';
import { Arrow } from '../Arrow/Arrow.component';

interface PositionContainerProps  extends React.HTMLAttributes<HTMLDivElement> {
    active?: boolean;
    arrowIndent?: string;
    arrowSize?: string;
    background?: string;
    border?: string;
    borderRadius?: string;
    children?: any;
    className?: string;
    color?: string;
    containerHeight: number;
    containerWidth: number;
    delay?: string;
    height: number;
    spacing?: number;
    maxWidth?: string;
    offset?: string;
    padding?: string;
    position: Position;
    shadow?: string;
    showArrow?: boolean;
    width: number;
    wrapContent?: boolean;
}

const StyledPositionContainer = styled('div')<PositionContainerProps>`
    position: absolute;
    box-sizing: border-box;
    display: flex;
    z-index: 1;
    min-width: 10rem;
    opacity: 0;
    visibility: hidden;
    transition: opacity 250ms ease-in-out;
    font-family: base;

    ${({
        border,
        borderRadius,
        color,
        delay,
        padding,
        shadow,
    }) => css({
        border,
        color,
        padding,
        borderRadius,
        boxShadow: shadow,
        transitionDelay: delay,
    })};

    ${({
        containerHeight,
        containerWidth,
        height,
        spacing,
        position,
        width,
    }) =>
        positionVariants(
            position,
            height,
            width,
            containerHeight,
            containerWidth,
            spacing // This is the distance of the item from the target in px
        )};

    &.active {
        opacity: 1;
        visibility: visible;
    }

    ${({ wrapContent = true }) =>
        css({ whiteSpace: wrapContent ? 'normal' : 'nowrap' })};

    ${({ maxWidth = 'auto' }) => css({ width: maxWidth })};
`;

interface BackgroundProps {
    background?: string;
}

const Background = styled('div')<BackgroundProps>`
    position: absolute;
    box-sizing: border-box;
    border-radius: inherit;
    z-index: 0;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;

    ${({ background }) =>
        css({
            background: th.color(background),
        })};
`;

const StyledOverlay = styled('div')`
    position: relative;
    z-index: 1;
`;

export const PositionContainer = forwardRef(
    (
        {
            active,
            arrowIndent,
            arrowSize,
            background = 'black',
            border,
            borderRadius = 'base',
            children,
            className,
            color,
            containerHeight,
            containerWidth,
            delay,
            height,
            spacing = 0,
            padding = '0.5rem',
            position = 'bottom',
            shadow,
            showArrow,
            width,
        }: PositionContainerProps,
        ref: React.RefObject<any>
    ) => (
        <StyledPositionContainer
            border={border}
            borderRadius={borderRadius}
            className={classNames(className || 'anchor-position-container', active && 'active')}
            color={color}
            containerHeight={containerHeight}
            containerWidth={containerWidth}
            delay={delay}
            height={height}
            spacing={spacing}
            padding={padding}
            position={position}
            ref={ref}
            shadow={shadow}
            width={width}
        >
            <Background background={background} />

            <StyledOverlay>
                {children}
            </StyledOverlay>

            {showArrow && (
                <Arrow
                    position={position}
                    size={arrowSize}
                    indent={arrowIndent}
                    background={background}
                    shadow={shadow}
                    border={border}
                />
            )}
        </StyledPositionContainer>
    )
);
