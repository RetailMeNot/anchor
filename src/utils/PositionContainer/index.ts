export { PositionContainer } from './PositionContainer.component';
