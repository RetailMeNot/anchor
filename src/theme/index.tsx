// Theme Files
import { typography } from './typography.theme';
import { ColorsTheme } from './colors.theme';
import { ThemeProvider as XstyledThemeProvider } from '@xstyled/styled-components';
import styled from '@xstyled/styled-components';
import { space as spaceStyles, SpaceProps } from '@xstyled/system';
import { BUTTON_KEY, BUTTON_THEME } from '../Button';
import { INPUT_KEY, INPUT_THEME } from '../Form/Input/utils';
import React from 'react';
import {Button} from '../Button';
import classNames from 'classnames';

export const RootTheme = {
    typography,
    radii: { none: '0', base: '4px', modal: '8px', circular: '1000px' },
    [BUTTON_KEY]: BUTTON_THEME,
    [INPUT_KEY]: INPUT_THEME,
    colors: ColorsTheme,
    fonts: {
        base:
            'Avenir Next, Segoe UI, Helvetica Neue, Helvetica, Roboto, sans-serif',
        mono: `'SFMono-Regular', Consolas, Liberation Mono, Menlo, Courier, monospace`,
    },
    // These are all minimums
    breakpoints: {
        xs: 0,
        sm: 576,
        md: 768,
        lg: 992,
        xl: 1200,
        xxl: 1600,
    },
    // todo: we should likely remove the color.borders colors and just use this
    // or find a way to reference the colors.border colors here
    // e.g. `thin solid borders.base`
    borders: {
        light: `thin solid ${ColorsTheme.borders.light}`,
        base: `thin solid ${ColorsTheme.borders.base}`,
        dark: `thin solid ${ColorsTheme.borders.dark}`,
    },

    skeleton: {
        colorStart: '#E7E7E7',
        colorEnd: '#D3D3D3',
    },
};

// Exporting out @xstyled's ThemeProvider. Only way I could get it to work w/o issues.
// export class ThemeProvider2 extends XstyledThemeProvider {}

interface IterableObject {
    [s: number]: number | string | boolean | IterableObject;
  }

interface DebugProps {
    property: number | string | boolean | IterableObject;
    propertyName: string;
    rootProperty?: boolean;
}

interface StyledDebugProps {
    indent?: number;
}

const StyledDebug = styled('div')<StyledDebugProps>`
    font-family: mono;
    margin-top: 0.5rem;
    color: white;

    .category .category h2 {
        background-color: #ec8947;
    }

    .category .category .box {
        border-color: #ec8947;
    }

    .category .category .category h2 {
        background-color: #b84343;
        color: white;
    }

    .category .category .category .box {
        border-color: #b84343;
    }

    .category .category .category .category h2 {
        background-color: #91cdd2;
        color: black;
    }

    .category .category .category .category .box {
        border-color: #91cdd2;
    }

    .category .category .category .category .category h2 {
        background-color: #567224;
        color: white;
    }

    .category .category .category .category .category .box {
        border-color: #567224;
    }
`;

const StyledH2 = styled('h2')`
    ${spaceStyles}
    background-color: #ffea92;
    color: black;
    margin-bottom: 0;
    margin-top: 0;
    padding: 0.25rem 0.5rem 0.25rem 1rem;
    font-size: 1rem;
    cursor: pointer;
    border-radius: 0 1rem 1rem 0;
`;

interface BoxProps extends SpaceProps {
    show?: boolean;
}

const Box = styled('div')<BoxProps>`
    ${spaceStyles}
    width: 20rem;
    border-left: solid #ffea92 2px;
    margin: 0.5rem;
    ${({show}) => `display: ${show ? 'block' : 'none'}`}
`;

const Category = styled('div')<SpaceProps>`
    ${spaceStyles}
    width: 20rem;
`;


const ColorBox = styled('div')<any>`
    width: 1rem;
    height: 1rem;
    display: inline-block;
    vertical-align: middle;
    margin-left: 0.5rem;
    background-color: ${({backgroundColor}) => backgroundColor};
    border: solid white 1px;
`;

const Debug = ({property, propertyName}: DebugProps) => {
    const [show, setShow] = React.useState(false);

    return(
        <StyledDebug>
            {(property || property === 0)
                ? (
                    typeof property === 'number' ||
                    typeof property === 'string' ||
                    typeof property === 'boolean' ||
                    typeof property === 'function'
                        ? (
                                <>
                                    <span style={{fontWeight: 'bold'}}>{propertyName}: </span>
                                    <span>
                                        {typeof property === 'function'
                                            ? 'function'
                                            : /^#[a-zA-Z0-9]{4,7}/.test(property.toString())
                                                ? (<>
                                                    <span style={{fontStyle: 'italic'}}>{property.toString()}</span>
                                                    <ColorBox backgroundColor={property} />
                                                    </>)
                                                : <span style={{fontStyle: 'italic'}}>{property.toString()}</span>
                                        }
                                    </span>
                                </>
                            )
                        :
                            <Category pl={0} className="category">
                                <StyledH2 onClick={() => setShow(!show)}>{propertyName}</StyledH2>
                                <Box pl={2} className="box" show={show}>
                                    {Object.values(property).map((_property, index) => (
                                        <Debug
                                            key={index}
                                            property={_property}
                                            propertyName={Object.getOwnPropertyNames(property)[index]}
                                        />
                                    ))}
                                </Box>
                            </Category>
                    )
                : (
                    <>
                       Empty.
                    </>
                )
            }
        </StyledDebug>
    );
};

const DebugWrapper = styled('div')`
    background-color: rgba(0,0,0,.75);
    border-left-color: transparent;
    position: fixed;
    left: 0;
    top: 0;
    height: 100%;
    overflow: hidden;
    width: 30rem;
    padding: 1rem 2rem 1rem 0;

    &.closed {
        height:2.4rem;
        width: auto;
        padding-right: 1rem;

        .debug-scroll {
            display: none;
        }
    }
`;

const DebugScroll = styled('div')`
    height: 100%;
    overflow-y: auto;
    width: 100%;
`;

interface ThemeProviderProps {
    children?: any;
    debug?: boolean;
    theme?: any;
}

export const ThemeProvider = ({ children, debug, theme }: ThemeProviderProps) => {
    const isOpen = localStorage.themeProviderDebug === 'true';
    const [isOpenState, setIsOpenState] = React.useState(isOpen);

    return (
        <XstyledThemeProvider theme={theme}>
            {debug &&
                <DebugWrapper className={classNames(!isOpenState ? 'closed' : null)}>
                    <Button
                        onClick={() => {
                            localStorage.themeProviderDebug = !isOpenState;
                            setIsOpenState(!isOpenState);
                        }}
                        size="sm"
                    >
                        ThemeProvider Debug
                    </Button>

                    <DebugScroll className="debug-scroll">
                        <Debug property={theme} propertyName="Theme" />
                    </DebugScroll>
                </DebugWrapper>
            }
            {children}
        </XstyledThemeProvider>
    );
};
