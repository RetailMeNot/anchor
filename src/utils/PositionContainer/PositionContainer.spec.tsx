// VENDOR
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { ThemeProvider } from '@xstyled/styled-components';
import { shallow, mount } from 'enzyme';
// ANCHOR
import { RootTheme } from '../../../src/theme';
import { PositionContainer } from './PositionContainer.component';

describe('Component: PositionContainer', () => {
    it('should be defined.', () => {
        const subject = (
            <ThemeProvider theme={RootTheme}>
                <PositionContainer
                    width={100}
                    height={50}
                    containerHeight={120}
                    containerWidth={70}
                    position="bottom"
                 />
            </ThemeProvider>
        );
        const wrapper = mount(subject);
        const component = shallow(subject);

        expect(subject).toBeDefined();
        expect(wrapper).toBeDefined();
        expect(component).toBeDefined();
        const tree = renderer.create(subject).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
