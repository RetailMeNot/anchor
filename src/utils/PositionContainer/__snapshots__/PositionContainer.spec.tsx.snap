// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: PositionContainer should be defined. 1`] = `
.c0 {
  position: absolute;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  z-index: 1;
  min-width: 10rem;
  opacity: 0;
  visibility: hidden;
  -webkit-transition: visibility 250ms ease-in-out,opacity 250ms ease-in-out;
  transition: visibility 250ms ease-in-out,opacity 250ms ease-in-out;
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  padding: 0.5rem;
  border-radius: 4px;
  left: 0;
  top: 0;
  -webkit-transform: translate3d(15px,50px,0);
  -ms-transform: translate3d(15px,50px,0);
  transform: translate3d(15px,50px,0);
  white-space: normal;
  width: auto;
}

.c0.active {
  opacity: 1;
  visibility: visible;
}

.c1 {
  position: absolute;
  box-sizing: border-box;
  border-radius: inherit;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background: black;
}

.c2 {
  position: relative;
  z-index: 1;
  width: 100%;
}

<div
  className="c0 anchor-position-container"
  height={50}
  spacing={0}
  width={100}
>
  <div
    className="c1 anchor-position-bg"
  />
  <div
    className="c2 anchor-position-overlay"
  />
</div>
`;
